From 3d0633da76595ff9720d544d460422fc49250f1c Mon Sep 17 00:00:00 2001
From: Nickey Yang <nickey.yang@rock-chips.com>
Date: Wed, 21 Nov 2018 15:58:06 +0800
Subject: [PATCH] Revert "libweston-desktop/xdg_shell_v6: Send error on
 wrongly-sized buffer"

this patch revert
https://lists.freedesktop.org/archives/wayland-devel/2017-March/033425.html
reconfigure weston_desktop_xdg_surface toplevel base.

Signed-off-by: Nickey Yang <nickey.yang@rock-chips.com>
---
 libweston-desktop/xdg-shell-v6.c | 39 +++++++++++++-------------------
 1 file changed, 16 insertions(+), 23 deletions(-)

diff --git a/libweston-desktop/xdg-shell-v6.c b/libweston-desktop/xdg-shell-v6.c
index 10274e0..741d6c7 100644
--- a/libweston-desktop/xdg-shell-v6.c
+++ b/libweston-desktop/xdg-shell-v6.c
@@ -636,6 +636,7 @@ weston_desktop_xdg_toplevel_committed(struct weston_desktop_xdg_toplevel *toplev
 {
 	struct weston_surface *wsurface =
 		weston_desktop_surface_get_surface(toplevel->base.desktop_surface);
+	bool reconfigure = false;
 
 	if (!wsurface->buffer_ref.buffer && !toplevel->added) {
 		weston_desktop_xdg_toplevel_ensure_added(toplevel);
@@ -644,30 +645,22 @@ weston_desktop_xdg_toplevel_committed(struct weston_desktop_xdg_toplevel *toplev
 	if (!wsurface->buffer_ref.buffer)
 		return;
 
-	struct weston_geometry geometry =
-		weston_desktop_surface_get_geometry(toplevel->base.desktop_surface);
-
-	if ((toplevel->next.state.maximized || toplevel->next.state.fullscreen) &&
-	    (toplevel->next.size.width != geometry.width ||
-	     toplevel->next.size.height != geometry.height)) {
-		struct weston_desktop_client *client =
-			weston_desktop_surface_get_client(toplevel->base.desktop_surface);
-		struct wl_resource *client_resource =
-			weston_desktop_client_get_resource(client);
-
-		wl_resource_post_error(client_resource,
-				       ZXDG_SHELL_V6_ERROR_INVALID_SURFACE_STATE,
-				       "xdg_surface buffer does not match the configured state");
-		return;
+	if (toplevel->next.state.maximized || toplevel->next.state.fullscreen)
+		reconfigure =
+			((toplevel->next.size.width != wsurface->width) ||
+			(toplevel->next.size.height != wsurface->height));
+
+	if (reconfigure)
+		weston_desktop_xdg_surface_schedule_configure(&toplevel->base, false);
+	else {
+		toplevel->current.state = toplevel->next.state;
+		toplevel->current.min_size = toplevel->next.min_size;
+		toplevel->current.max_size = toplevel->next.max_size;
+
+		weston_desktop_api_committed(toplevel->base.desktop,
+					     toplevel->base.desktop_surface,
+					     sx, sy);
 	}
-
-	toplevel->current.state = toplevel->next.state;
-	toplevel->current.min_size = toplevel->next.min_size;
-	toplevel->current.max_size = toplevel->next.max_size;
-
-	weston_desktop_api_committed(toplevel->base.desktop,
-				     toplevel->base.desktop_surface,
-				     sx, sy);
 }
 
 static void
-- 
2.17.1

